fix(sdk/middleware): support `apply` endpoint for `orderEdits`

adnasa opened the pull request:
#### Summary

-  support `apply` endpoint for `orderEdits`
- closes #2058

#### Description

In `sdk/middleware`, we `build()` request [for the respective service](https://github.com/commercetools/nodejs/blob/master/packages/api-request-builder/src/default-services.js#L1:L1) in `sdk`.

It turned out that the `service.build()` is coupled to an option for `orderEdits` that determines whether an `order/edits/${id}/apply` endpoint should be build.

Here is how we reason when building the `order/edits/${id}/apply` endpoint in `sdk/middleware`:

- GIVEN that `service=orderEdits` and `applyOrderEditTo` has a value: [we build the apply endpoint](https://docs.commercetools.com/api/projects/order-edits#apply-an-orderedit)
- GIVEN that `service=orderEdits` and `applyOrderEditTo` has no value: [we build the update endpoint](https://docs.commercetools.com/api/projects/order-edits#update-orderedit-by-id)

adnasa commented on packages/sdk/src/middleware/middleware.ts:
```diff
@@ -37,7 +37,17 @@ const actionToUri = (action: TSdkAction, projectKey?: string): string => {
+    // given `service=orderEdits` and no `applyOrderEditTo`, we build an update endpoint
+    // https://docs.commercetools.com/api/projects/order-edits
+    applyOrderEdit:
+      action.payload.service === 'orderEdits' &&
```
This is a bit hacky, but the request-builder doesn't expose any type that I can make use of.

adnasa commented on packages/sdk/src/middleware/middleware.ts:
```diff
@@ -37,7 +38,16 @@ const actionToUri = (action: TSdkAction, projectKey?: string): string => {
+    // https://docs.commercetools.com/api/projects/order-edits
+    applyOrderEdit:
+      action.payload.service === SUPPORTED_SERVICES.orderEdits &&
+      typeof action.payload.options?.applyOrderEditTo === 'string',
```
we want to ensure that it is a string, not `null` nor a number

qmateub replied:
There is one thing that I'm not sure I follow 100%. Is this parameter then going to be send for all the services and not just for `order-edits` (to `false` obviously) or is this part of the code only related to the interaction with the `order-edits` endpoint?

adnasa replied:
This will always be sent to `service.build()`
https://github.com/commercetools/nodejs/blob/master/packages/api-request-builder/src/create-service.js#L128

Which by default is `false`.
We continue sending false as long as it doesn't fulfil the requirements discussed in the PR description

qmateub commented on packages/sdk/src/middleware/middleware.ts:
```diff
@@ -37,7 +38,16 @@ const actionToUri = (action: TSdkAction, projectKey?: string): string => {
+      action.payload.service === SUPPORTED_SERVICES.ORDER_EDITS &&
+      typeof action.payload.options?.applyOrderEditTo === 'string',
+    // at this stage, the `projectKey` should be available already.
+    withProjectKey: true,
```
Why it was not there before? 🤔

adnasa replied:
because it is defaulted https://github.com/commercetools/nodejs/blob/master/packages/api-request-builder/src/create-service.js#L128

I added it here to be more explicit.

qmateub reviewed (commented):
Got what the problem is 👍 . I think I'm maybe missing some context 😅

mohib0306 reviewed (approved):
code looks good to me. I was wondering if we need to specify all the supported services now. I see that only one is used and the rest are just defined but not used as of now

tdeekens commented on packages/sdk/src/middleware/middleware.ts:
```diff
@@ -37,7 +38,16 @@ const actionToUri = (action: TSdkAction, projectKey?: string): string => {
+    // given `service=orderEdits` and given `applyOrderEditTo`, we build an apply endpoint
+    // given `service=orderEdits` and no `applyOrderEditTo`, we build an update endpoint
+    // https://docs.commercetools.com/api/projects/order-edits
+    applyOrderEdit:
+      action.payload.service === SUPPORTED_SERVICES.ORDER_EDITS &&
+      typeof action.payload.options?.applyOrderEditTo === 'string',
```
```suggestion
    applyOrderEdit: getShouldApplyOrderEdit(action)
```

Somewhere above:

```js
function getShouldApplyOrderEdit(action) {
   return (
      action.payload.service === SUPPORTED_SERVICES.ORDER_EDITS &&
      typeof action.payload.options?.applyOrderEditTo === 'string'
   );
}
```

Generally I wonder what would happen if we demand to pass it in as an option on the payload? So `action.payload.options.applyOrderEditTo`. Would that work or is it against the API?

adnasa replied:
no, we can not really demand it, unless we use unions.
But using unions wouldn't make that the much sense to implement since the consumer can't necessarily be TS

tdeekens replied:
Got yah. Thanks for clarifying. To combine with my point above. Code looks sharp, tests suberb. I'd subjectively not have the constants and slam the check into a small function.

adnasa commented on packages/sdk/src/constants.ts:
```diff
@@ -0,0 +1,44 @@
+const SUPPORTED_SERVICES = {
```
yeah we could.
I want to be verbose that `service` implies an endpoint.

tdeekens replied:
Yup, I can see that just wondering if there is a true benefit in this case. Having this "work-around" isolated might also be nice.

qmateub reviewed (approved):
🚀

adnasa commented:
@tdeekens I dropped the services `constants` file and extracted the string into a constant with reference to the original `default-services` in `nodejs`

tdeekens reviewed (approved):
Thanks!

adnasa commented:
these snapshots emits the following diff.

![image](https://user-images.githubusercontent.com/462507/108521146-b6754b80-72cb-11eb-8a4a-d4d049dbf40e.png)

I'm going to take a gamble here and approve them
